extensions/kinesis/src/main/java/com/hazelcast/jet/kinesis/KinesisSources.java

```diff
@@ -126,15 +131,16 @@ private KinesisSources() {
      * also to construct the source once configuration is done
      */
     @Nonnull
-    public static Builder kinesis(@Nonnull String stream) {
-        return new Builder(Objects.requireNonNull(stream));
+    public static Builder<Map.Entry<String, byte[]>> kinesis(@Nonnull String stream) {
+        return new Builder<>(Objects.requireNonNull(stream));
     }
 
     /**
      * Fluent builder for constructing the Kinesis source and setting
      * its configuration parameters.
+     * @param <T> Result type returned from the source.
      */
-    public static final class Builder {
+    public static final class Builder<T> {
```

Comment:
Could you make this Builder class implements Serializable? Since we had a reference to `KinesisSources$BuilderBuilder#toArray(Record, Shard)` in the default `projectionFn`, we get a serialization exception in the serialization of `PipelineImpl` when trying to use the default kinesis source. See the stack trace (with -Dsun.io.serialization.extendedDebugInfo=true) for the details:
```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.jet.impl.pipeline.PipelineImpl'

	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:129)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:240)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:216)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:201)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:156)
	at com.hazelcast.jet.impl.AbstractJobProxy.doSubmitJob(AbstractJobProxy.java:317)
	at com.hazelcast.jet.impl.AbstractJobProxy.<init>(AbstractJobProxy.java:104)
	at com.hazelcast.jet.impl.JobProxy.<init>(JobProxy.java:65)
	at com.hazelcast.jet.impl.JetInstanceImpl.newJobProxy(JetInstanceImpl.java:155)
	at com.hazelcast.jet.impl.AbstractJetInstance.newJobInt(AbstractJetInstance.java:144)
	at com.hazelcast.jet.impl.AbstractJetInstance.newJob(AbstractJetInstance.java:126)
	at com.hazelcast.jet.JetService.newJob(JetService.java:63)
	at com.hazelcast.jet.kinesis.KinesisFailureTest.networkOutageWhileRunning(KinesisFailureTest.java:134)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:115)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:107)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.NotSerializableException: com.hazelcast.jet.kinesis.KinesisSources$Builder
	- element of array (index: 0)
	- array (class "[Ljava.lang.Object;", size: 5)
	- field (class "java.lang.invoke.SerializedLambda", name: "capturedArgs", type: "class [Ljava.lang.Object;")
	- object (class "java.lang.invoke.SerializedLambda", SerializedLambda[capturingClass=class com.hazelcast.jet.kinesis.KinesisSources$Builder, functionalInterfaceMethod=com/hazelcast/function/FunctionEx.applyEx:(Ljava/lang/Object;)Ljava/lang/Object;, implementation=invokeSpecial com/hazelcast/jet/kinesis/KinesisSources$Builder.lambda$build$98db9676$1:(Lcom/hazelcast/jet/kinesis/impl/AwsConfig;Ljava/lang/String;Lcom/hazelcast/jet/retry/RetryStrategy;Lcom/hazelcast/jet/kinesis/impl/source/InitialShardIterators;Lcom/hazelcast/jet/core/EventTimePolicy;)Lcom/hazelcast/jet/core/ProcessorMetaSupplier;, instantiatedMethodType=(Lcom/hazelcast/jet/core/EventTimePolicy;)Lcom/hazelcast/jet/core/ProcessorMetaSupplier;, numCaptured=5])
	- custom writeObject data (class "com.hazelcast.jet.impl.pipeline.transform.StreamSourceTransform")
	- object (class "com.hazelcast.jet.impl.pipeline.transform.StreamSourceTransform", Kinesis Source (TestStream))
	- custom writeObject data (class "java.util.HashMap")
	- object (class "java.util.LinkedHashMap", {Kinesis Source (TestStream)=[map], map=[mapWithMergingSink(Results)], mapWithMergingSink(Results)=[]})
	- field (class "com.hazelcast.jet.impl.pipeline.PipelineImpl", name: "adjacencyMap", type: "interface java.util.Map")
	- root object (class "com.hazelcast.jet.impl.pipeline.PipelineImpl", Pipeline {Kinesis Source (TestStream)=[map], map=[mapWithMergingSink(Results)], mapWithMergingSink(Results)=[]})
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1183)
	at java.base/java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1379)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1175)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:349)
	at com.hazelcast.jet.impl.pipeline.transform.StreamSourceTransform.writeObject(StreamSourceTransform.java:137)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1145)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1497)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:349)
	at java.base/java.util.LinkedHashMap.internalWriteEntries(LinkedHashMap.java:332)
	at java.base/java.util.HashMap.writeObject(HashMap.java:1412)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1145)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1497)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1553)
	at java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1510)
	at java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1433)
	at java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1179)
	at java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:349)
	at com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:124)
	at com.hazelcast.internal.serialization.impl.defaultserializers.JavaDefaultSerializers$JavaSerializer.write(JavaDefaultSerializers.java:117)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.write(StreamSerializerAdapter.java:39)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:237)
	... 23 more
```

Reply:
We previously added a serialization version field,
`private static final long serialVersionUID = 1L;`, to all serializable classes for making them more adaptable to implementation changes in https://github.com/hazelcast/hazelcast/pull/19534. But, it seems that we forgot to add this field to the `KinesisSources` here. Could you also add this field to `com.hazelcast.jet.kinesis.KinesisSources` and `com.hazelcast.jet.kinesis.KinesisSources$Builder` after making it implements Serializable?

Reply:
We prefer not to make the builders `Serializable`. To work around the field capture we store the field into a local variable of in the build method. It is done for the other fields, but not for the new one.